source/language/pulses.rst

```diff
@@ -92,3 +79,217 @@ circuit. Most of the time symbols in the ``defcal`` table will also have
 corresponding ``gate`` definitions. However, if a user provides a ``defcal`` for a symbol
 without a corresponding ``gate``, then we treat such operations like the ``opaque`` gates
 of prior versions of OpenQASM.
+
+OpenPulse instructions
+======================
+
+In addition to OpenQASM instructions, ``defcal`` blocks may contain OpenPulse 
+instructions using a text-based version of OpenPulse defined here. OpenPulse is 
+a JSON format described in the paper 
+`Qiskit Backend Specifications for OpenQASM and OpenPulse Experiment <https://arxiv.org/abs/1809.03452>`_.
+
+The text format described here has several advantages over the equivalent JSON 
+format:
+
+- It is more readable
+- Absolute time is handled through built-in timekeeping instructions
+- Gates and classical instructions can be mixed in with the pulses to create far richer calibrations
+
+Channels
+--------
+
+Channels (sometimes referred to as "frames") describe the different forms of
+classically-controlled stimulus fields used to interact with a qubit. There are
+two kinds of channels: input channels, which send signal into the system to
+drive the qubit, and output channels, which measure signal from the system to
+measure qubits.
+
+Each channel has a user-specified local oscillator frequency which is assumed to
+have been specified outside of the context of OpenQASM.
+
+A value of type ``channel`` can be retrieved with the ``drive`` (input channels) 
+and ``acquire`` (output channels) functions, a reference to the physical qubit, 
+and the name of the channel. If the name of the channel is omitted then the
+names "drive" and "acquire" will be used respectively.
+
+.. code-block:: none
+
+   // Get the drive input channel for qubit 0
+   drive(%0)
+   // Get the acquire output channel for qubit 1
+   acquire(%1)
+   // Get an input channel named "measure" for qubit 1
+   // Could be used to specify the measurement stimulus channel
+   drive(%1, "measure")
+   // Get a custom input channel named "cr1" for qubit 0
+   drive(%0, "cr1")
+
+Play instruction
+----------------
+
+Pulses are scheduled using the ``play`` instruction. Play instructions have two
+required parameters:
+- the channel on which to play the pulse
+- an array of complex samples which define the amplitude points for the pulse envelope
+
+The length of the pulse will be the length of the array multiplied by the unit
+``dt``, which specifies the sample rate.
+
+.. code-block:: none
+
+   // Play a 3 sample pulse on qubit 0's drive channel
+   play drive(%0), [1+0*j, 0+1*j, 1/sqrt(2)+1/sqrt(2)*j];
+
+Specifying a full list of samples for real-life pulses can be unwieldy, so we
+include several built-in pulse shape functions as well:
```

Comment:
Seems like a worthwhile design space to explore. This would require two things:
- a `shape` type that is not immediately turned into an array of samples on creation
- an addition of a length parameter on the `play` command which right now is implicit